src/Rule/Boolean.php

```diff
@@ -26,32 +33,62 @@ final class Boolean implements RuleWithOptionsInterface, SkipOnEmptyInterface, S
+     *
+     * - https://www.php.net/manual/en/language.operators.comparison.php
+     * - https://www.php.net/manual/en/types.comparisons.php
+     * - https://www.php.net/manual/en/language.types.type-juggling.php
+     *
+     * Defaults to `false` meaning non-strict mode is used.
+     * @param string $messageWithType Error message used when validation fails and neither non-scalar value (int,
+     * float, string, bool) nor null was provided as input. The type is used instead of value here for more predictable
+     * formatting.
+     *
+     * You may use the following placeholders in the message:
+     *
+     * - `{attribute}`: the label of the attribute being validated.
+     * - `{true}` - the value set in {@see $trueValue} option.
+     * - `{false}` - the value set in {@see $falseValue} option.
+     * - `{type}`: the type of the value being validated.
+     * @param string $messageWithValue Error message used when validation fails and either scalar value (int, float,
+     * string, bool) or null was provided as input.
+     *
+     * You may use the following placeholders in the message:
+     *
+     * - `{attribute}`: the label of the attribute being validated.
+     * - `{true}` - the value set in {@see $trueValue} option.
+     * - `{false}` - the value set in {@see $falseValue} option.
+     * - `{value}`: the value being validated.
+     * @param bool|callable|null $skipOnEmpty Whether to skip this rule if the validated value is empty / not passed.
+     * See {@see SkipOnEmptyInterface}.
+     * @param bool $skipOnError Whether to skip this rule if any of the previous rules gave an error. See
+     * {@see SkipOnErrorInterface}.
+     * @param Closure|null $when A callable to define a condition for applying the rule. See {@see WhenInterface}.
+     * @psalm-param WhenType $when
+     */
     public function __construct(
-        /**
-         * @var scalar the value representing "true" status. Defaults to `1`.
-         */
         private int|float|string|bool $trueValue = '1',
-        /**
-         * @var scalar the value representing "false" status. Defaults to `0`.
-         */
         private int|float|string|bool $falseValue = '0',
-        /**
-         * @var bool whether the comparison to {@see $trueValue} and {@see $falseValue} is strict. When this is `true`,
-         * the value and type must both match those of {@see $trueValue} or {@see $falseValue}. Defaults to `false`,
-         * meaning only the value needs to be matched.
-         */
         private bool $strict = false,
-        private string $nonScalarMessage = 'Value must be either "{true}" or "{false}".',
-        private string $scalarMessage = 'Value must be either "{true}" or "{false}".',
-
-        /**
-         * @var bool|callable|null
-         */
-        private $skipOnEmpty = null,
+        private string $messageWithType = self::DEFAULT_MESSAGE,
+        private string $messageWithValue = self::DEFAULT_MESSAGE,
```

Comment:
Is it really necessary to separate this cases?

Reply:
They were already separated, just additionally renamed, because handling of `null` was added. Or do you mean something related to using constant?

The reason of separation was predictable formatting of complex structures. The tests show that.